rust/osm2lanes/src/transform/tags_to_lanes/bus.rs

```diff
@@ -55,84 +70,101 @@ pub(super) fn bus(
-        );
-    }
-    if tags.is(BUSWAY + locale.driving_side.opposite().tag(), "lane") {
-        if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
-            road.forward_inside_mut()
-                .ok_or_else(|| RoadError::unsupported_str("no forward lanes for busway"))?
-                .set_bus(locale)?;
-        } else {
-            return Err(
-                RoadMsg::ambiguous_str("busway:BACKWARD=lane for bidirectional roads").into(),
+        if tags.is(BUSWAY + "both", "lane") {
+            busway.forward_direction = Infer::Direct(Some(Direction::Forward));
+            busway.backward_direction = Infer::Direct(Some(Direction::Backward));
+            if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
+                warnings.push(RoadMsg::ambiguous_str("busway:both=lane for oneway roads").into());
+            }
+        }
+        if tags.is(BUSWAY + locale.driving_side.tag(), "lane") {
+            busway.forward_direction = Infer::Direct(Some(Direction::Forward));
+        }
+        if tags.is(BUSWAY + locale.driving_side.tag(), "opposite_lane") {
+            warnings.push(
+                RoadMsg::ambiguous_tag(BUSWAY + locale.driving_side.tag(), "opposite_lane").into(),
             );
         }
+        if tags.is(BUSWAY + locale.driving_side.opposite().tag(), "lane") {
+            if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
+                busway.forward_direction = Infer::Direct(Some(Direction::Forward));
+            } else {
+                warnings.push(
+                    RoadMsg::ambiguous_str("busway:BACKWARD=lane for bidirectional roads").into(),
+                );
+            }
+        }
+        if tags.is(
+            BUSWAY + locale.driving_side.opposite().tag(),
+            "opposite_lane",
+        ) {
+            if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
+                // TODO: does it make sense to have a backward lane on the forward_side????
+                busway.forward_direction = Infer::Direct(Some(Direction::Backward));
+            } else {
+                warnings.push(RoadMsg::Ambiguous {
+                    description: None,
+                    tags: Some(tags.subset(&[
+                        BUSWAY + locale.driving_side.opposite().tag(),
+                        TagKey::from("oneway"),
+                        TagKey::from("oneway:bus"),
+                    ])),
+                });
+            }
+        }
+
+        busway
     }
-    if tags.is(
-        BUSWAY + locale.driving_side.opposite().tag(),
-        "opposite_lane",
-    ) {
-        if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
-            // TODO: does it make sense to have a backward lane on the forward_side????
-            let lane = road
-                .forward_inside_mut()
+}
+
+impl RoadBuilder {
+    pub fn set_busway_scheme(
```

Comment:
Yeah, this two-step approach of "figure out what's being described" and then "apply it to the road being built up" is way cleaner to read (and even unit test more granularly, if we ever wanted to)